### PR TITLE
Obsolete Encoding.UTF7 property and UTF7Encoding ctors

### LINKED ISSUE #32284: Proposal: Obsolete UTF-7 encoding in the framework
## Motivation

The UTF-7 transcoding logic in the framework is problematic. UTF-7 itself has been deprecated for quite some time, and MSDN [even warns against](https://docs.microsoft.com/en-us/dotnet/api/system.text.utf7encoding) using the `UTF7Encoding` type due to the possibility of introducing security and reliability problems.

ASP.NET Full Framework 4.5+ and ASP.NET Core (all versions) forbid processing UTF-7 encoded requests by default due to the problems that UTF-7 has historically caused.

Ideally we'd be able to change application code not to consider UTF-7 as a default candidate for text processing, but that's not always practical. There are scenarios where library code passes untrusted _charset_ values as inputs to the `Encoding.GetEncoding` API, and they're not necessarily hardened against these APIs returning `UTF7Encoding` or some other undesirable instance. A proposal which properly deprecates UTF-7 framework-wide also needs to take this into account.

## Proposed API and behavioral changes

```cs
namespace System.Text
{
    public class Encoding
    {
        [Obsolete(...)] // ** NEW attribute ** on existing property
        public static Encoding UTF7 { get; }
    }

    public class UTF7Encoding
    {
        [Obsolete(...)] // ** NEW attribute ** on existing ctor
        public UTF7Encoding();

        [Obsolete(...)] // ** NEW attribute ** on existing ctor
        public UTF7Encoding(bool);
    }
}
```

 * Add `[Obsolete]` (as warning) to the `System.Text.UTF7Encoding` ctors; this allows things like `typeof(UTF7Encoding)` to continue to work
 * Add `[Obsolete]` (as warning) to the static property `System.Text.Encoding.UTF7`
 * Disallow `System.Text.Encoding.GetEncoding("utf-7")` from returning a `UTF7Encoding` instance by default

If an application needs `Encoding.GetEncoding("utf-7")` to return a `UTF7Encoding` instance, this can be done by using `AppContext` at app start, re-introducing UTF-7 knowledge into the system, as shown below.

```cs
AppContext.SetSwitch("System.Text.Encoding.EnableUnsafeUTF7Encoding", true);
```

### PR DESCRIPTION
**Pull request by GrabYourPitchforks**
Resolves https://github.com/dotnet/runtime/issues/32284.

### Breaking changes

 * The property `Encoding.UTF7` is now obsolete as warning (diagnostic id _BCL0001_).
 * The `UTF7Encoding` constructors are now obsolete as warning (diagnostic id _BCL0001_). The type itself is not obsolete in order to allow code like `if (x is UTF7Encoding)` to continue to work without warning.
 * The APIs `Encoding.GetEncoding(65000 /* UTF-7 code page */)` and `Encoding.GetEncoding("utf-7")` will now throw an "unknown code page" exception.

`AppContext` can be used to re-enable `Encoding.GetEncoding` recognizing the string "utf-7" or the code page 65000.

This resulted in some changes to unit tests. /cc @carlossanlop @jozkee @dotnet/ncl to review their areas of ownership and to make sure I didn't torpedo the tests. For the _System.Net.Http.Json_ UTF-7 tests, I removed the test body entirely rather than suppressing, as the point of this change was to disallow `HttpClient` and friends from honoring `Content-Type: ...; charset=utf-7` without explicit opt-in by the app developer.

/cc @terrajobst as well since this our first usage of the "better obsoletion" APIs that were added, so we should settle on URL and diagnostic id prefix.

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Obsoletions.cs
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Obsoletions.cs:9]: Why not use regular CamelCase here?

File: src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchive.cs
Comment (stephentoub) [inline src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchive.cs:378]: Was this already dead code because we weren't setting FEATURE_UTF7?  Are we setting FEATURE_UTF32?

File: src/libraries/System.IO.Ports/tests/SerialPort/Encoding.cs
Comment (stephentoub) [inline src/libraries/System.IO.Ports/tests/SerialPort/Encoding.cs:114]: Do obsoletion IDs need to be numerical?  If not, have we considered using more descriptive names that would avoid the need for a comment?  May not be a great idea to break from convention even if it's possible.

File: src/libraries/System.IO.Ports/tests/SerialPort/ReadTo.cs
Comment (stephentoub) [inline src/libraries/System.IO.Ports/tests/SerialPort/ReadTo.cs:29]: Can we use our standard casing for consts?

File: src/libraries/System.IO.Ports/tests/SerialPort/Read_char_int_int.cs
Comment (stephentoub) [inline src/libraries/System.IO.Ports/tests/SerialPort/Read_char_int_int.cs:38]: Utf7WebName

File: src/libraries/System.Private.CoreLib/src/System/Text/Encoding.cs
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Text/Encoding.cs:244]: What custom provider?

File: src/libraries/System.Private.CoreLib/src/System/Text/EncodingTable.cs
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Text/EncodingTable.cs:116]: Should we instead just ensure it's not in s_mappedCodePages?  Then anything else using it won't accidentally include it, and we won't pay the List overhead on every GetEncodings call.

**Comment by Tratcher** [comment]
> /cc @jkotalik and @Tratcher since there's some aspnet code which touches the `Encoding.UTF7` property

All AspNetCore usages I see are either checks to prevent it from being used, or  tests for explicit usage. We may have to tweak the code a bit, but there shouldn't be any functional changes.

File: src/libraries/System.IO.Ports/tests/SerialPort/ReadTo.cs
Comment (GrabYourPitchforks) [reply src/libraries/System.IO.Ports/tests/SerialPort/ReadTo.cs:29]: I'm following the convention already present in this file.

File: src/libraries/System.IO.Ports/tests/SerialPort/Read_char_int_int.cs
Comment (GrabYourPitchforks) [reply src/libraries/System.IO.Ports/tests/SerialPort/Read_char_int_int.cs:38]: I'm following the convention already present in this file.

File: src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchive.cs
Comment (GrabYourPitchforks) [reply src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchive.cs:378]: Correct, neither of these is set by the build process. Would you prefer I remove both?

File: src/libraries/System.Private.CoreLib/src/System/Text/Encoding.cs
Comment (GrabYourPitchforks) [reply src/libraries/System.Private.CoreLib/src/System/Text/Encoding.cs:244]: If you call [`Encoding.RegisterProvider`](https://docs.microsoft.com/en-us/dotnet/api/system.text.encoding.registerprovider), you can override the return value of `Encoding.GetEncoding("utf-7")`, `Encoding.GetEncoding("utf-8")`, and any other encoding instance. If you register a custom provider for UTF-7, we never query the `AppContext` switch because this switch statement never get hit.

Registering a custom provider doesn't change the return value of the built-in accelerators (like `Encoding.UTF7` or `Encoding.UTF8`) since they're hardcoded to always use the built-in implementation.

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Text/Encoding.cs:244]: Right.  It's the "re-enabled" lingo that confuses me.  There's no way for someone to get at our built-in Encoding to re-enable it via Encoding.RegisterProvider, is there?

File: src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchive.cs
Comment (stephentoub) [reply src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchive.cs:378]: If it's dead code, yeah.

Comment (tarekgh) [reply src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchive.cs:378]: I looked at the generated IL from System.IO.Compression and seeing neither UTF7 nor UTF32 is defined. either we need to confirm these are intentional not including them and then remove the code for UTF32 too, or we enable such ifdef's (not enabling UTF7 though)



 private set
 {
  if (value != null && (value.Equals(Encoding.BigEndianUnicode) || value.Equals(Encoding.Unicode)))
  {
   throw new ArgumentException(SR.EntryNameEncodingNotSupported, "EntryNameEncoding");
  }
  _entryNameEncoding = value;
 }

---
In reply to: [436265058](https://github.com/dotnet/runtime/pull/37535#discussion_r436265058) [](ancestors = 436265058)

File: src/libraries/System.IO.Ports/tests/SerialPort/Read_char_int_int.cs
Comment (tarekgh) [reply src/libraries/System.IO.Ports/tests/SerialPort/Read_char_int_int.cs:38]: I would say for this one, we don't have to be consistent with the rest of the file and Use the Camel case better.

---
In reply to: [436877176](https://github.com/dotnet/runtime/pull/37535#discussion_r436877176) [](ancestors = 436877176)

File: src/libraries/System.IO.Ports/tests/SerialPort/ReadTo.cs
Comment (tarekgh) [reply src/libraries/System.IO.Ports/tests/SerialPort/ReadTo.cs:29]: I am seeing this constant value defined in other file too. can we define it in one place and reference it from there instead of defining it twice? 

---
In reply to: [436877102](https://github.com/dotnet/runtime/pull/37535#discussion_r436877102) [](ancestors = 436877102)

File: src/libraries/System.IO.Ports/tests/SerialPort/ReadTo.cs
Comment (tarekgh) [inline src/libraries/System.IO.Ports/tests/SerialPort/ReadTo.cs:697]: >com1.Encoding.WebName == UTF7_WEBNAME; [](start = 34, length = 38)

I would prefer comparing the CodePage instead of comparing names.

File: src/libraries/System.IO.Ports/tests/SerialPort/ReadTo.cs
Comment (tarekgh) [inline src/libraries/System.IO.Ports/tests/SerialPort/ReadTo.cs:867]: >BCL0001 [](start = 24, length = 7)

I am seeing you are disabling the warning in 2 places in this project. can we cache Encoding.UTF7 in one place and disable the warning there and then reference the cached values else where?

File: src/libraries/System.Private.CoreLib/src/System/Text/Encoding.cs
Comment (tarekgh) [inline src/libraries/System.Private.CoreLib/src/System/Text/Encoding.cs:79]: >ENABLEUTF7_APPCONTEXT_SWITCH_NAME [](start = 29, length = 33)

Can we define this in the file [https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.CoreLib/src/System/LocalAppContextSwitches.cs](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.CoreLib/src/System/LocalAppContextSwitches.cs)

just to be easy to find all config switches in one place

File: src/libraries/System.Private.CoreLib/src/System/Text/Encoding.cs
Comment (tarekgh) [inline src/libraries/System.Private.CoreLib/src/System/Text/Encoding.cs:1044]: >IsUTF7EncodingEnabled [](start = 29, length = 21)

please move this to [https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.CoreLib/src/System/LocalAppContextSwitches.cs](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.CoreLib/src/System/LocalAppContextSwitches.cs)

File: src/libraries/System.Private.CoreLib/src/System/Text/EncodingTable.cs
Comment (tarekgh) [inline src/libraries/System.Private.CoreLib/src/System/Text/EncodingTable.cs:109]: >encodingInfos [](start = 31, length = 13)

can we just use array as before and decide about the length according to the config switch value? this will help avoid allocating the List.

File: src/libraries/System.Private.CoreLib/src/System/Text/EncodingTable.cs
Comment (tarekgh) [inline src/libraries/System.Private.CoreLib/src/System/Text/EncodingTable.cs:107]: >GetEncodings [](start = 39, length = 12)

After my change enabling the encoding enumeration through the providers, I am wondering should we allow UTF7 provided by the provider? I would say not allowing it except if the config switch is turned on.

Comment (GrabYourPitchforks) [reply src/libraries/System.Private.CoreLib/src/System/Text/EncodingTable.cs:107]: If somebody went out of their way to create a custom provider that implements UTF-7, why would we disable it? They've already opted-in by virtue of registering the provider.

For context, when I spoke offline with Shawn the recommendation was that _all_ non-Unicode encodings should be disabled by default. This means that arbitrary code page encodings would be disabled by default as well. And it turns out this is already the case, since unless you register the code page provider, the .NET runtime doesn't know anything about how to resolve those encodings.

Comment (tarekgh) [reply src/libraries/System.Private.CoreLib/src/System/Text/EncodingTable.cs:107]: > If somebody went out of their way to create a custom provider that implements UTF-7, why would we disable it? They've already opted-in by virtue of registering the provide

I believe enabling or disabling the UTF-7 decision should be taken by the app (through the config switch) and not any other place. Imagine a library using a provider that has UTF-7 (even the library not caring about this encoding) and then the app using such a library. That means the library took the decision on behalf of the app which I think is not right. Another point is consistency and easy to understand the behavior regardless of there are providers or not.

> when I spoke offline with Shawn the recommendation was that all non-Unicode encodings should be disabled by default.

This is proven impossible as we saw a lot of scenarios that require non-Unicode encoding. like networking protocols using ISO encodings for instance. I know Shawn passionate for Unicode which I share that too but the reality is really different.

Comment (GrabYourPitchforks) [reply src/libraries/System.Private.CoreLib/src/System/Text/EncodingTable.cs:107]: Of course. I mentally appended "plus ASCII and Latin-1" to what he said, even if those exact words never came from his mouth. 😉

Comment (tarekgh) [reply src/libraries/System.Private.CoreLib/src/System/Text/EncodingTable.cs:107]: :-)

Also, .NET has to be compliant and support specific encodings (e.g. GB13080 for instance). disabling these by default can create unwanted issues :-)

File: src/libraries/System.Net.Http.Json/tests/UnitTests/TranscodingReadStreamTests.cs
Comment (GrabYourPitchforks) [reply src/libraries/System.Net.Http.Json/tests/UnitTests/TranscodingReadStreamTests.cs:240]: To be honest, I debated this. In the end I asked myself two questions:

> Does the library intend to provide special support for UTF-7? Does this test provide additional coverage beyond what is already covered by the existing tests for UTF-16, UTF-32, ASCII, and Latin-1?

Since I believe the answer to these questions is no I felt the test could be safely deleted. But if there's a counterargument to be made this is definitely the right forum! :)

Comment (Jozkee) [reply src/libraries/System.Net.Http.Json/tests/UnitTests/TranscodingReadStreamTests.cs:240]: IMO the `Transcoding*` classes are most intended for UTF-16 -> UTF-8 and vice versa and also as you mention, we are no losing coverage so I think is fine to delete the tests. Thanks for explaining.

File: src/libraries/System.Private.CoreLib/src/System/Text/EncodingTable.cs
Comment (GrabYourPitchforks) [reply src/libraries/System.Private.CoreLib/src/System/Text/EncodingTable.cs:116]: Changing that static backing field ends up being fairly intrusive because so many different things depend on it (including the contents of other backing fields), and it results in us having to special-case UTF-7 throughout more methods like `InternalGetCodePageDataItem`. I'll drop the use of `List<T>` here.

File: src/libraries/System.IO.Ports/tests/SerialPort/ReadTo.cs
Comment (GrabYourPitchforks) [reply src/libraries/System.IO.Ports/tests/SerialPort/ReadTo.cs:697]: Sure. I'll flow this change throughout the project, and it should also address other reviewers' concerns about adding new consts to these files.

File: src/libraries/System.IO.Ports/tests/SerialPort/Read_char_int_int.cs
Comment (GrabYourPitchforks) [reply src/libraries/System.IO.Ports/tests/SerialPort/Read_char_int_int.cs:38]: This field no longer exists as of newest iteration.

File: src/libraries/System.IO.Ports/tests/SerialPort/ReadTo.cs
Comment (GrabYourPitchforks) [reply src/libraries/System.IO.Ports/tests/SerialPort/ReadTo.cs:29]: This field no longer exists as of newest iteration.

**Comment by jkotas** [comment]
@eerhardt Anything we want to adjust here to make it easy for the linker to strip the UTF7Encoding?

File: src/libraries/System.Private.CoreLib/src/System/LocalAppContextSwitches.cs
Comment (eerhardt) [inline src/libraries/System.Private.CoreLib/src/System/LocalAppContextSwitches.cs:15]: Would it make sense to unify on a single naming convention with these AppContext settings?

I know the ones in this file use the pattern `Switch.<namespace>.<name>`. But we also have others that don't start with `Switch` - for example

https://github.com/dotnet/runtime/blob/bd6cbe3642f51d70839912a6a666e5de747ad581/src/coreclr/src/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.cs#L13-L17

And I'm adding a new one for EventSource:

https://github.com/dotnet/runtime/pull/38129/files#diff-f5b0a60692bde26a95ea15bb86a6bfdbR239

```C#
        private static bool InitializeIsSupported() =>
            AppContext.TryGetSwitch("System.Diagnostics.Tracing.EventSource.IsSupported", out bool isSupported) ? isSupported : true;
```

And we plan on adding one for the `System.SR.UsingResourceKeys`, https://github.com/dotnet/runtime/issues/34057, and a DebugMode setting, https://github.com/dotnet/runtime/pull/37288.

I know we can't rename any setting we've already shipped. But going forward, it would be good to have a standard naming convention for new switches (and possibly creating alternative switch names for the existing non-conforming names).

ASP.NET appears to not use the `Switch.` prefix:
https://github.com/dotnet/aspnetcore/blob/6afd6f40c043f96c8fb1ee06cbebbc1d229efddc/src/Identity/Core/src/SignInManager.cs#L380
https://github.com/dotnet/aspnetcore/blob/6afd6f40c043f96c8fb1ee06cbebbc1d229efddc/src/Http/Routing/src/Matching/DfaMatcherBuilder.cs#L44

**Comment by eerhardt** [comment]
> Anything we want to adjust here to make it easy for the linker to strip the UTF7Encoding?

Yes, as it exists today the linker won't be able to strip UTF7Encoding since it is being rooted by `System.Text.Encoding::GetEncoding(System.Int32)`.

To enable stripping, we can make the new switch a full blown [feature switch](https://github.com/dotnet/designs/blob/master/accepted/2020/feature-switch.md). You can see an example of this in https://github.com/dotnet/runtime/pull/38129.

Basically the thing we will want to add is an entry into this file
https://github.com/dotnet/runtime/blob/38aa0b76a6423b902cc4b6c310cad9c49ba675b6/src/libraries/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.Shared.xml#L2-L6

that tells the linker when `LocalAppContextSwitches.EnableUnsafeUTF7Encoding` property can be stubbed out as `return false;`. This will allow any code that branches based on the property to get the branch removed during linking time.

Something like:

```xml
    <type fullname="System.LocalAppContextSwitches">
      <method signature="System.Boolean get_EnableUnsafeUTF7Encoding()" body="stub" value="false" feature="Switch.System.Text.Encoding.EnableUnsafeUTF7Encoding" featurevalue="false" />
    </type>
```

File: src/libraries/System.Private.CoreLib/src/System/LocalAppContextSwitches.cs
Comment (tarekgh) [reply src/libraries/System.Private.CoreLib/src/System/LocalAppContextSwitches.cs:15]: I agree not using `Switch` prefix. it is not adding any value and also the new switches we are introducing are not using this prefix anymore.

**Comment by GrabYourPitchforks** [comment]
Thanks @eerhardt for the tip! I'll plumb it through in the next iteration.

File: src/libraries/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.Shared.xml
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.Shared.xml:7]: This looks reversed to me. I believe that this would allow UTF7Encoding to be unconditionally enabled; but not trimmed.

Comment (GrabYourPitchforks) [reply src/libraries/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.Shared.xml:7]: I think you're right. Even though the default value is _false_, we still want trimming to take place if somebody has set _false_.

**Comment by GrabYourPitchforks** [comment]
@jkotas any idea why the wasm tests might be failing?
https://helix.dot.net/api/2019-06-17/jobs/b400b19c-d5d5-41c2-a0a5-10064c66ac86/workitems/System.Text.Encoding.Tests/console

```txt
[23:20:06] dbug: test[0]
      [FAIL] System.Text.Encodings.Tests.EncodingMiscTests.NormalizationTest(codepage: 65000, normalized: False, normalizedC: False, normalizedD: False, normalizedKC: False, normalizedKD: False)
      
[23:20:06] dbug: test[0]
      System.NotSupportedException : Support for UTF-7 is disabled. See https://aka.ms/dotnet-warnings/MSLIB0001 for more information.
```

There is a _runtimeconfig.template.json_ file present in that test directory which re-enables the UTF-7 code paths, so the test project succeeds on all other platforms. See https://github.com/dotnet/runtime/blob/9d6811c5753d0d627d608a4d5d932aee014f43e5/src/libraries/System.Text.Encoding/tests/runtimeconfig.template.json.

**Comment by jkotas** [comment]
> @jkotas any idea why the wasm tests might be failing?

I do not think wasm supports .runtimeconfig.json files. @safern ?

**Comment by akoeplinger** [comment]
> > @jkotas any idea why the wasm tests might be failing?

> I do not think wasm supports .runtimeconfig.json files. @safern ?

It's not supported right now, but I think we should make it work.

I've opened https://github.com/dotnet/runtime/issues/38433 to track that, you can add an `[ActiveIssue(..., TestPlatforms.Browser)]` on the affected tests.

**Comment by GrabYourPitchforks** [comment]
Awesome - thanks both for the pointer! :)

**Comment by GrabYourPitchforks** [comment]
Test failure is https://github.com/dotnet/runtime/issues/38605. I'm just waiting for a review signoff before merging.

@tarekgh I think I addressed all your feedback?

**Review by tarekgh — APPROVED** [review]
Thanks @GrabYourPitchforks. LGTM.

**Comment by tarekgh** [comment]
@GrabYourPitchforks just checking, did you log and fill a breaking change template for this one? sorry if I missed it.

**Comment by GrabYourPitchforks** [comment]
@tarekgh Not yet, thanks for the reminder. :)
I'll add it to the docs repo and make sure the aka.ms link forwards to it.

**Comment by GrabYourPitchforks** [comment]
Breaking change notification: https://github.com/dotnet/docs/issues/19274

**Pull request merged**
